### ISSUE TITLE
Struggling with equally named templates

### ISSUE THREAD

**Issue by ctron** [la10736/rstest#124]
I am trying to use `rstest-reuse` and it works great. Until I added a second template, with the same name, in a different module of my crate.

It looks like it is creating a macro which gets exported, and thus "pollutes" the root namespace of the crate.

So I tried to define the template in a central location, and re-use it from different sub-modules. Also without success as the `apply` macro panics, as it expects an `Ident` as argument.

Importing the template into the module's namespace using `use` also doesn't work:

~~~
error: macro-expanded `macro_export` macros from the current crate cannot be referred to by absolute paths
  --> src/tests/mqtt/mod.rs:17:5
   |
17 | use crate::mqtt_integration_versions;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[deny(macro_expanded_macro_exports_accessed_by_absolute_paths)]` on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #52234 <https://github.com/rust-lang/rust/issues/52234>
~~~

I am not sure what the right approach here is.

**Comment by ctron** [comment]
Dropping the `macro_export` attribute make the problem go away. And I can have multiple templates with the same name in different sub-modules.

**Comment by la10736** [comment]
Ok, `#![macro_export]` do exactly that. I've documented this behavior in the Readme notes:

> If you plan to spread your templates in some modules and you use a different name for each template consider to add the 
> global attribute !#[macro_use] at crate level: this put all your templates available everywhere: you should just take care that 
> a template should be defined before the apply call.

For the first part of your question I wrote in the same Readme

> If you define a template in a module and you want to use it outside the module you should lift it by mark the module with the 
> `#[macro_use]` attribute. This attribute make your template visible outside this module but not at the upper level. When a
> template is defined you can use it in all submodules that follow the definition.

I don't know what I can do more :(

To understand better: when you say _Dropping the `macro_export` attribute_ you means drop it from your crate... It's right?

**Comment by ctron** [comment]
Indeed that is documented when you look more closely. I was focusing this section here: https://github.com/la10736/rstest#use-parametrize-definition-in-more-tests

Regarding dropping the `macro_export` take a look at the following example:

~~~rust
mod foo {
    mod bar {
        use rstest_reuse::{self, *};

        #[template]
        #[rstest(b, case(1))]
        pub fn my_template(#[case] v: i32) {}
    }
    mod baz {
        use rstest_reuse::{self, *};

        #[template]
        #[rstest(b, case(1))]
        pub fn my_template(#[case] v: i32) {}
    }
}
~~~

This will fail with the following error:

~~~
error[E0428]: the name `my_template` is defined multiple times
  --> src/main.rs:16:9
   |
9  |         #[template]
   |         ----------- previous definition of the macro `my_template` here
...
16 |         #[template]
   |         ^^^^^^^^^^^ `my_template` redefined here
   |
   = note: `my_template` must be defined only once in the macro namespace of this module
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
~~~

Which, as you say is expected. But it is only because the macros generated by the `template` macro are exported as well. Which may not be required, depending in your use case.

I made the following change to the template macro (I parsing the argument can probably be done in a better way):

https://github.com/ctron/rstest/blob/e866168f79c52cb33dae6253c0df75af4aa5ffe0/rstest_reuse/src/lib.rs#L204-L232

This now only adds the "export" if "local" is not added. Switching the example above the following now makes it work:

~~~rust
mod foo {
    mod bar {
        use rstest_reuse::{self, *};

        #[template(local)]
        #[rstest(b, case(1))]
        pub fn my_template(#[case] v: i32) {}
    }
    mod baz {
        use rstest_reuse::{self, *};

        #[template(local)]
        #[rstest(b, case(1))]
        pub fn my_template(#[case] v: i32) {}
    }
}
~~~

I personally would prefer adding `(export)` for templates that should be exported. But keeping it compatible makes a lot of sense too.

**Comment by la10736** [comment]
Thank. Now I understand what you mean.

Your idea is correct... Just one thing: what about to use `#[local]` attribute instead?

```
mod foo {
    mod bar {
        use rstest_reuse::{self, *};

        #[template]
        #[local]
        #[rstest(b, case(1))]
        pub fn my_template(#[case] v: i32) {}
    }
    mod baz {
        use rstest_reuse::{self, *};

        #[template]
        #[local]
        #[rstest(b, case(1))]
        pub fn my_template(#[case] v: i32) {}
    }
}
```
You are right `#[export]` would be better... We can introduce both of them and raise a warning for default that will change at the next release.

**Comment by ctron** [comment]
True, that would work as well. I will try to implement this and update the PR. Macros like that are still uncharted territory for me :smile: , but I like the challenge!

**Comment by la10736** [comment]
Ok I decided for a braking change. We should use `#[export]` to export and nothing to local. 

Emit a warning that introduce the change will force everybody to change the code to shut up the warning and and no export is the is the best as default behavior.

**Comment by la10736** [comment]
What is missed:
- [ ] impl that catch e remove `#[export]`
- [ ] Tests
- [ ] Update readme
- [ ] Update docs
- [ ] fill changelog

**Comment by DeveloperC286** [comment]
@la10736 @ctron Just opened a pull request adding a unit test for the current local template implementation.

#131

**Comment by la10736** [comment]
@DeveloperC286 Thanks... I'll push my changes ASAP.

**Comment by la10736** [comment]
Sorry  for the delay... I've released now the 0.2.0 version that make `#[export]` optional and the standard behavior become don't use  `#[macro_export]` annotation
